skbio/sequence/_grammared_sequence.py

```diff
@@ -809,12 +808,21 @@ def to_definites(self, degenerate="wild", noncanonical=True):
                     f"Invalid character '{degenerate}' in sequence. Character must "
                     f"be within sequence alphabet: {self.alphabet}"
                 )
+        elif isinstance(degenerate, str) and len(degenerate) == 0:
+            sub_char = 0
         else:
             raise ValueError('Invalid value for parameter "degenerate".')
 
-        sequence = "".join(sub_char if x in degenerates else x for x in sequence)
+        htab = np.zeros((256,), dtype=np.uint8)
```

Comment:
The hash table should be stored as a hidden attribute of the class (like `_degenerate_codes`), rather than being created every time one calls the `to_definites` method. This could save lots of compute. Because there are two versions: 1) degenerate codes, 2) degenerate and noncanonical codes, you will need to create two attributes (e.g., `_degenerate_hash` and `_degen_nonca_hash`).